src/main/java/frc/robot/subsystems/MecanumDriveSubsystem.java

```diff
@@ -79,28 +101,212 @@ public Map<String,RelativeEncoder> GetEncoders()
+
+      driveBase.driveCartesian(chassisSpeeds.vxMetersPerSecond, chassisSpeeds.vyMetersPerSecond,
+          chassisSpeeds.omegaRadiansPerSecond);
+
+    } else {
+      driveBase.driveCartesian(
+          deadzone(forwardSpeed, Constants.Misc.driveDeadzone) * driveGain,
+          deadzone(rightSpeed, Constants.Misc.driveDeadzone) * driveGain,
+          deadzone(rotatinalSpeed, Constants.Misc.driveDeadzone) * rotationGain);
+    }
+  }
+
+  public void toggleFieldOrient() {
+    m_fieldOriented = !m_fieldOriented;
+    System.out.printf("field orient: %b%n", m_fieldOriented);
+  }
+
+  public void sportMode(boolean setting) {
+    sportMode = setting;
+  }
+
+  public void alignAmp() {
+    int targetId = 0;
+    Optional<Alliance> ally = DriverStation.getAlliance();
+    if (ally.isPresent()) {
+      if (ally.get() == Alliance.Red) {
+        targetId = 5;
+      }
+      if (ally.get() == Alliance.Blue) {
+        targetId = 6;
+      }
+    }
+    else {
+      //Doesn't do anything/exits out of the function
+      return;
+    }
+    Pose3d pose3d = atfl.getTagPose(targetId).get();
+
+    double targetX = pose3d.getX();
+    double targetY = pose3d.getY();
+    //double desiredHeading = 270 + 180 + Constants.Alignment.ampAngleOffsetDegrees;
+    double desiredRange = 1.2;
+    //alignTarget(desiredx, desiredy, desiredHeading);
+    SmartDashboard.putNumber("Tx", targetX);
+    SmartDashboard.putNumber("Ty", targetY);
+    SmartDashboard.putString("Target","Amp");
+    alignrange(desiredRange, targetX, targetY);
+  }
+
+  public void alignSpeaker() {
+    int targetIdSpeaker = 0;
+    Optional<Alliance> ally = DriverStation.getAlliance();
+    if (ally.isPresent()) {
+      if (ally.get() == Alliance.Red) {
+        targetIdSpeaker = 4;
+      }
+      if (ally.get() == Alliance.Blue) {
+        targetIdSpeaker = 7;
+      }
+    }
+    else {
+      //Doesn't do anything/exits out of the function
+      return;
+    }
+    Pose3d pose3d = atfl.getTagPose(targetIdSpeaker).get();
+
+    double targetX = pose3d.getX();
+    double targetY = pose3d.getY();
+    //double desiredHeading = 270 + 180 + Constants.Alignment.ampAngleOffsetDegrees;
+    double desiredRange = 1.5;
+    //alignTarget(desiredx, desiredy, desiredHeading);
+    SmartDashboard.putNumber("Tx", targetX);
+    SmartDashboard.putNumber("Ty", targetY);
+    SmartDashboard.putString("Target","Speaker");
+    alignrange(desiredRange, targetX, targetY);
+  }
+  
+  public void alignrange(double desiredRange, double targetX, double targetY) {
+    Pose2d robotPose = m_LimelightNavigation.getPose2d();
+    double currentHeading = robotPose.getRotation().getDegrees();
+
+    double desiredHeading = Math.toDegrees(Math.atan2(targetY, targetX));
+    double headingError = desiredHeading-currentHeading;
+
+    // Normalizes heading error because we had an issue with aligning.
+
+    if (headingError > 180) {
```

Comment:
consider making this a method (lower priority)

Reply:
donezi